code/PipeFlowMain.py

```diff
@@ -64,3 +97,17 @@ def pipeSize(L, Q, hl, e):
         pipeChoice += 1
     return 0
 
+
+# Determines required pipe size given Pipe length L [m], flowrate Q [m^3/s], headloss hl [m] and roughness [mm]
+# Returns exact pipe diameter in meters
+def exactPipeSize(L,Q,hl,e):
+    pipeDiameter1 = 0.5  # Roughly the midpoint of schedule 40 pipe sizes
+    while True:
+        V = (4 * Q) / (math.pi * pipeDiameter1 ** 2)
+        f = getFrictionFactor(V, pipeDiameter1, e)
+        pipeDiameter2 = ((8*L*Q**2*f)/(math.pi**2*g*hl))**0.2
+        if pipeDiameter2 == pipeDiameter1:
+            return pipeDiameter2
+        else:
+            pipeDiameter1 = pipeDiameter2
+
```

Comment:
Could you rewrite to ensure this function terminates. Two floats will never be identically equal. So change it to (D_old -D_new)^2<epsilon (small number).

Reply:
I changed the equality to use the math.isclose method, Thanks for the heads up, I had forgotten about floating point errors